### ISSUE TITLE
Investigate Hearings Management Queue Results

### ISSUE THREAD

**Issue by pshahVA** [department-of-veterans-affairs/caseflow#11882]
## Hearings management queue may be limiting results

- The hearings management queue is only returning about 200 tasks, but there were thousands prior to temporarily limiting the volume for performance reason.

- Some engineering work has been done to close admin tasks that are corrected, but we still need to check if all tasks are showing up.

**Comment by pkarman** [comment]
There is not a numerical limit. There is a filter on AOD tasks:

https://github.com/department-of-veterans-affairs/caseflow/blob/master/app/controllers/organizations/tasks_controller.rb#L24

**Comment by pshahVA** [comment]
Thanks.  @lomaxap said he updated the code so that many tasks may have been closed,  we may be able to remove this filter now.  He said he will look.

**Comment by rubaiyat22** [comment]
@lomaxap @pshahVA 

I did a benchmark test on retrieving all tasks of a current user of Hearing Management organization.
``` 
RequestStore[:current_user] = User.system_user
org = HearingsManagement.singleton
timing = Benchmark.measure { GenericQueue.new(user: org).tasks }
```
Task count: 5546

Run 1: 
```
=> #<Benchmark::Tms:0x000055663a2d8598 @label="", @real=18.471204397999827, @cstime=0.0, @cutime=0.0, @stime=0.41777, @utime=12.132155000000001, @total=12.549925000000002>
```
Run 2:
```
=> #<Benchmark::Tms:0x00005566340b9790 @label="", @real=17.060635664000074, @cstime=0.0, @cutime=0.0, @stime=0.40796600000000005, @utime=11.025085, @total=11.433051>
```
Run 3:
```
=> #<Benchmark::Tms:0x000055663555bc48 @label="", @real=18.687685814999895, @cstime=0.0, @cutime=0.0, @stime=0.33956600000000003, @utime=12.767992, @total=13.107558>
```
Real value is the elapsed time in minutes.

On average of **3** runs, **5546** tasks took **18.0731752923** minutes.

**Comment by pshahVA** [comment]
Does this mean that the volume coming back on the UI is too small?

**Comment by ferristseng** [comment]
@pshahVA Are you talking about with the fix in place? I think we're returning 1000 right now. Do you think we should be returning more?

We were trying to judge how removing the fix would affect prod performance. It looks like there might still be timeout issues if we do revert it.

**Comment by pshahVA** [comment]
Yes. When I showed andrew, we only see like 300 on the UI. I wanted to make sure nothing was missing.

**Comment by lomaxap** [comment]
It looks like we're expecting children tasks to be [eagerly loaded](https://github.com/department-of-veterans-affairs/caseflow/blob/master/app/models/queues/generic_queue.rb#L49), so [this](https://github.com/department-of-veterans-affairs/caseflow/blob/master/app/models/task.rb#L151-L153) should not be an n+1 method. But my console output console out put is clearly showing that it is.

**Comment by rubaiyat22** [comment]
Just for clarity, what is a n+1 method?

and I just tried the following code and got this result
```
timing = Benchmark.measure {tasks = Task.incomplete_or_recently_closed.where(assigned_to: org).where(appeal_type: [Appeal.name, LegacyAppeal.name]).includes({appeal: [:available_hearing_locations, :claimants]}, :assigned_by, :assigned_to, :children, :parent).order(created_at: :asc).all.to_a}
=> #<Benchmark::Tms:0x000055694226ee30 @label="", @real=2.5041814840005827, @cstime=0.0, @cutime=0.0, @stime=0.0882960000000006, @utime=2.080909000000002, @total=2.1692050000000025>
```
```
tasks.count
=> 4951
```
It does seem that finding the *TimedHoldTask* child is taking the most time.

**Comment by lomaxap** [comment]
> Just for clarity, what is a n+1 method?
[this](https://medium.com/@bretdoucette/n-1-queries-and-how-to-avoid-them-a12f02345be5) is a pretty good explanation. Basically any query where there's an additional query for each record that's returned. Eagerly loading related tables with `#includes` allows us to get all related records at once instead of one at a time.

**Comment by ferristseng** [comment]
I was able to figure out which attributes that trigged database calls in the AMA task serializer were not necessary to how the organization tasks page functioned. The cost savings we observed was about 3x (from 30s to 10s for ~600 tasks).

We don't know the impact of how this would affect other parts of the code, but could look into it.

Here are the attributes with expensive calls we can remove:

```
assigned_to.name
placed_on_hold_at
on_hold_duration
aod
available_hearing_locations
document_id
available_actions
```
